packages/jaeger-ui/src/components/TracePage/CriticalPath/index.tsx

```diff
@@ -78,12 +78,19 @@ function criticalPathForTrace(trace: Trace) {
   const rootSpanId = trace.spans[0].spanID;
   // If there is root span then algorithm implements
   if (rootSpanId) {
-    const spanMap = trace.spans.reduce((map, span) => {
-      map.set(span.spanID, span);
-      return map;
-    }, new Map<string, Span>());
+    // Use the pre-built spanMap
+    const spanMap = trace.spanMap;
+
+    // Populate childSpanIds from childSpans for critical path computation
+    // Critical path utilities need string[] not Span[] for manipulation
+    const spanMapWithChildIds = new Map<string, Span>();
+    spanMap.forEach((span, spanId) => {
+      const childSpanIds = span.childSpans.map(child => child.spanID);
+      spanMapWithChildIds.set(spanId, { ...span, childSpanIds });
+    });
+
     try {
-      const refinedSpanMap = getChildOfSpans(spanMap);
+      const refinedSpanMap = getChildOfSpans(spanMapWithChildIds);
```

Comment:
instead of transforming the top-level spanMap (where each span contains its children already) into another map with child IDs, we can just change the getChildOfSpans functions to work off the richer map